docs/proposals/20200615-external-remediation.md

```diff
@@ -0,0 +1,299 @@
+         * [Alternative 2 - Utilize conditions](#alternative-2---utilize-conditions)
+         * [Alternative 3 - Create a dedicated MHC](#alternative-3---create-a-dedicated-mhc)
+      * [Upgrade Strategy](#upgrade-strategy)
+      * [Additional Details](#additional-details)
+         * [Test Plan [optional]](#test-plan-optional)
+         * [Graduation Criteria [optional]](#graduation-criteria-optional)
+         * [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+      * [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Enable opt-in external remediation for unhealthy nodes backed by machines.
+
+## Motivation
+
+The Cluster API includes an optional Machine Healthcheck Controller component that implements automated health checking capability, however it doesn’t offer any other remediation than replacing the underlying infrastructures. 
+
+Environments consisting of hardware based clusters are significantly slower to (re)provision unhealthy machines, so they have a need for a remediation flow that includes at least one attempt at power-cycling unhealthy nodes.
+
+Other environments and vendors also have specific remediation requirements, such as KCP, so there is a need to provide a generic mechanism for implementing custom remediation logic.
+
+### Goals
+
+- Facilitate rapid experimentation by creating the ability to define customized remediation flows outside of the Machine Health Check and CAPI codebase.
+
+### Non-Goals/Future Work
+
+- Provide a [unified remediation flow](https://docs.google.com/document/d/1klvIR5wKbcCrl2vLPIbbWpmelGTDM-T0Qb4xw36P3zY/edit) as part of the MachineHealthCheck controller. See Alternative 1
+- Add power primitives to the Machine API.  While useful, it is unnecessary to the work proposed here and may already be covered by one or more of the following issues:
+    - Add Ability to Reboot Hosts
+    - Supporting power off/on
+    - Add machine/infra feature support fields
+- Alter the default deletion flow proposed in [cluster-api PR#3056](https://github.com/kubernetes-sigs/cluster-api/pull/3056)
+- Provide a finalizer-like pre-hook mechanism for allowing customization or blocking of the remediation process prior to power-cycling a machine or deleting it. This concept may already be covered as part of a separate enhancement
+- Create an escalation path from failed external remediation attempts to machine deletion.
+- Implement a formal mechanism for triggering recovery of workload and volumes by the k8s scheduler
+
+## Proposal
+
+We introduce a generic mechanism for supporting externally provided custom remediation strategies.
+
+We propose modifying the MachineHealthCheck CRD to support a remediationTemplate, an ObjectReference to a provider-specific template CRD. 
```

Comment:
Hmm, I believe that this relies on `generateTemplate(..)` which receives an `ObjectReference`.
Can we use `generateTemplate()` without having an `ObjectReference`?

Reply:
We could likely refactor the core logic of generateTemplate into a separate function that takes individual parameters (or an input struct) rather than an ObjectReference, calling that function from both net new code and still providing generateTemplate() for backward compatibility until we've removed existing use of ObjectReference.